challenges/Sudoku-BackTracking.md

```diff
@@ -0,0 +1,143 @@
+
+## C++ Code
+### NOTE: Only the snippets of the core functions required for proper functioning of the algorithm have been added. Also, in our sudoku grid, a '0' digit means an empty space.
+```cpp
+// declared as global variables for proper functioning
+int row;
+int col;
+
+bool findblank (int sudoku[9][9], int &row, int &col) //returns true if there is an unassigned value in grid
+{
+  // this function takes in a sudoku grid as an arguement 
+  // and returns true if there exist an empty space in our grid.
+  
+  //declare row and col as global variables initially for proper functionality
+  for (row = 0 ; row < 9; row++)
+  {
+    for (col = 0 ; col < 9; col++)
+    {
+      if (sudoku[row][col] == 0)
+      {
+        return true;
+      }
+    }
+  }
+  
+  return false;
+}
+
+bool checknum (int sudoku[9][9], int row, int col, int num)
+{
+  // function to check whether a given number is valid in the given row and column
+  
+  // iterates over all rows and reuturns false if the number already exist in the column.
+  for (int i = 0; i < 9; i++)
+  {
+    if (sudoku[i][col] == num)
+      return false;
+  }
+  
+  // iterates over all columns and reuturns false if the number already exist in the row.
+  for (int i = 0; i < 9; i++)
+  {
+    if (sudoku[row][i] == num)
+      return false;
+  }
+  
+  // checks the 3*3 box in which the number is to be placed
+  for (int i = row - row%3; i < row - row%3 + 3; i++)
+  {
+    for (int j = col - col%3; j < col - col%3 + 3; j++)
+    {   
+        if (sudoku[i][j] == num)
+          return false;
+    }
+  }
+  
+  // return false directly if the square is pre-occupied
+  if (sudoku[row][col] != 0)
+    return false;
+  
+  return true;
+}
+
+bool solvemain (int sudoku[9][9])
+{
+  // main function where backtracking algorithm has been implemented.
+
+  // if there is no empty square then the puzzle is solved hence return true.
+  // however if there is an empty square then the values of the global variables 
+  // row and column would have changed accordingly hence we make use of them
+  // in the following for-loop.
+  if (!findblank(sudoku, row, col)) // FIRST if condition
+    return true;
+  
+  // iterates over all the numbers and uses the modified values of row, col to 
+  // check whether the given number is valid in that position or not.
+  for (int num = 1; num < 10; num++)
+  {
+      // if the given number is valid then assign the number to that empty square
+      if (checknum (sudoku,row,col,num))
+      { 
+        sudoku[row][col] = num;
+        
+        // heart of the algorithm. 
+        // recursive call to solvemain.
+        // if the sudoku is completed then the last function will return true 
+        // from the FIRST if condition itself.
+        // hence the sudoku will be completed and we can return true.
+        if (solvemain(sudoku))
+			return true;
+          
+        // however if for some reason there comes a point when there is no valid
```

Comment:
Merge all these comments as one doc string

Reply:
Could you please elaborate? I'm not able to understand.

Reply:
@kaustubhgupta do you mean change the single line comments into multi-line comments?

Reply:
Yes, because currently is very messy